### DISCUSSION TITLE
Use table field `ToolTip` instead of page field `ToolTip` (runtime 13.0 / BC24)

### DISCUSSION THREAD

**Discussion by rvanbekkum** [StefanMaron/BusinessCentral.LinterCop#530] [New Rule]
**Scenario**
You have a have a `table` X and a `page` Y with `SourceTable = X` in the same app.
The page Y has a page field P that uses a table field T from table X as the source expression.

1. If page field P has a value for the `ToolTip` property and table field T does NOT have a value for the `ToolTip` property, then raise a diagnostic that informs the user that it is better to apply the tooltip on table field T in table object X.

2. If page field P has a value for the `ToolTip` property and table field T ALSO has a value for the `ToolTip` property but the value is exactly the same, then raise a diagnostic that informs the user about the redundant tooltip.

    Sidenote: Don't go as far as checking the translations. The original `ToolTip` value should not be the same.

3. If page field P has a value for the `ToolTip` property and table field T ALSO has a value for the `ToolTip` property but the value is NOT the same, then do not raise a diagnostic. Having a different tooltip is most likely deliberate (although we cannot know _for sure_).

The rules should only be applied when table and page are in the same app.

The rules should only be run and diagnostics should only be raised if runtime is 13.0 or higher (i.e., for BC 24.0 or newer).
Maybe Microsoft comes with something for this, so let's keep an eye out for that as well.

**Comment by rvanbekkum** [comment]
See also https://github.com/StefanMaron/BusinessCentral.LinterCop/discussions/490

**Comment by Arthurvdv** [comment]
Thank you for providing the details on how the rule should wordt, this was of great help to create the rule.

It's now available in the (pre)release version [v0.30.28](https://github.com/StefanMaron/BusinessCentral.LinterCop/releases/tag/v0.30.28) of the LinterCop.

**Reply by rvanbekkum** [reply]
It seems to work great!

Could we have one last improvement? :)
Raise a different diagnostic message if the table field tooltip is defined but the page field tooltip has the same value.

Currently the rule raises the message `Use table field ToolTip instead of page field ToolTip.` regardless of the scenario.

I think the most clear way to address this is _probably_ having two separate diagnostics:
1. Raise a completely different message for the scenario where table tooltip and page tooltip are the same, e.g.:
    `The ToolTip property of page field {0} and it's table field {1} have the same value and therefore the page field's ToolTip can be removed.`
2. Change the original message to something like:
    `A value for the ToolTip property is missing for the table field {0} of page field {1}. Consider adding a ToolTip for table field {0} and/or removing the ToolTip from page field {1}.`

What are your thoughts on this? Do you have other ideas?

**Reply by Arthurvdv** [reply]
You;re absolutely right, looking back this should indeed have been two separate rules.

I've split the rule into two now and used you're suggestions for the diagnostics, thank you for suggestion these! 🤗

|Id| Title|
|---|---|
|LC0064|Missing ToolTip property on table field.|Info|
|LC0066|Duplicate ToolTip between page and table field.|Info|
